### PR TITLE
[InstCombine] Support reassoc for foldLogicOfFCmps

### PR DESCRIPTION
**Pull request by nikic**
We currently support simple reassociation for foldAndOrOfICmps(). Support the same for foldLogicOfFCmps() by going through the common foldBooleanAndOr() helper.

This will also resolve the regression on #112704, which is also due to missing reassoc support.

I had to adjust one fold to add support for FMF flag preservation, otherwise there would be test regressions. There is a separate fold (reassociateFCmps) handling reassociation for *just* that specific case and it preserves FMF. Unfortuantely it's not rendered entirely redundant by this patch, because it handles one more level of reassociation as well.

### CODE REVIEW

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

Missing fold: https://alive2.llvm.org/ce/z/wawsAb
I will post a patch later.

**Comment by nikic** [comment]
Ping for the second opinion :)

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:1475]: You can do the and inline now
```suggestion
      Builder.setFastMathFlags(LHS->getFastMathFlags() & RHS->getFastMathFlags());
```

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:1476]: Really we ought to have FMF parameters for all the Create* functions instead of the flag guard boilerplate

**Review by arsenm — APPROVED** [review]

**Pull request merged**
